### DISCUSSION TITLE
ESP32-S3 with Xedge32: Some initial questions

### DISCUSSION THREAD

**Discussion by darethehair** [RealTimeLogic/Xedge32#28] [Q&A]
Today I picked up my mail and took home my first ESP32-S3, hooked it up to my Linux system, and configured the device for Xedge32 -- even trying a few lua commands. Cool!  

Even though I have a fair bit of experience running lua code on NodeMCU/EPS8266 devices (with no IDE and just using CLI to upload/testing!), this new environment is very new and strange to me :)

My hope was that I could easily upload my existing lua scripts and have everything run correctly out-of-the-box, I suspect from my initial lua command testing that the syntax (and even commands) will be somewhat different e.g. the command for connecting to WIFI.  If so, does any document exist to point out the lua version differences?  Is Xedge32 lua more powerful than regular NodeMCU lua?

Any other thoughts on my migration task are appreciated! :)

My primary ESP8266 lua project is a BME280 weather station sending data to Thingspeak, and also to HTML display using an internal simple webserver.  It works great, but I was forced to halt further development since I had run out of space/memory -- hence the desire to jump over to ESP32 -- as long as I could keep using lua.  I notice that the Xedge32 lua system includes the BME280 module, but none others -- is this going to be extended?

**Comment by surfskidude** [comment]
It's great to hear you got Xedge32 working. However,  while both NodeMCU and Xedge32 use Lua as their language, the APIs are very different. This difference is where the learning curve primarily lies. The core principles of Lua programming remain the same, which is a plus. However, the specific APIs provided by Xedge32 and the Barracuda App Server are unique and tailored to leverage the advanced capabilities of the ESP32 hardware with a focus on IoT.

I think it's a good idea to dedicate some time to familiarize yourself with these new APIs. The official Xedge32 documentation is an excellent starting point, offering detailed insights into the specifics of these APIs and how they differ from what you might be used to in the NodeMCU/ESP8266 environment. 

Regarding your query about the Lua version differences and extending the Lua system, these areas are often subject to continuous development and community contributions.

**Reply by darethehair** [reply]
Thanks for the response!  Today I started down the path you suggested, beginning with some simple HTTP 'GET' requests similar to what I have been using in my NodeMCU Lua code, using the syntax reference I have found here:

https://realtimelogic.com/ba/doc/en/lua/auxlua.html

I am immediately struck by how rich and complex the Xedge32 Lua language is -- almost too overwhelming for my simple requirements!  Strangely, some simple free API call tests I make appear to work, but others do not -- even though my example URL submissions work perfectly in my regular browser.  My inclination is to believe that it is my newbie ignorance that is probably the cause -- but I have now reached the point where I cannot progress unless someone familiar with the Xedge32 Lua sets me straight.  Is this the best forum to provide such basic help requests?

As an example, this HTML API 'GET' request works just fine in both my browser (just the URL) and my Lua code:

```
   local http=require"httpm".create()
   local t,err=http:json("http://api.sunrise-sunset.org/json",{})
   if t then -- If we got a Lua table (decoded JSON string)
      print("Sunrise:",t.results.sunrise)
   end
```
For today, I get this result:

`Sunrise:	5:58:31 AM`

However, this one -- with optional parameters -- does not:

```
   local http=require"httpm".create()
   local t,err=http:json("http://api.sunrise-sunset.org/json?lat=49&lng=-98&formatted=0",{})
   print(t,err)
   if t then -- If we got a Lua table (decoded JSON string)
      print("Sunrise:",t.results.sunrise)
   end
```

For this, today, I get this result:

```
nil	<html>
<head><title>404 Not Found</title></head>
<body bgcolor="white">
<center><h1>404 Not Found</h1></center>
<hr><center>nginx</center>
</body>
</html>

```

I have tested a bunch of other free APIs, retrieving IP addresses, etc. and the results are not totally consistent i.e. some work, some do not, even though the raw URLs work fine in my browser.  Different APIs fail in different ways -- almost as if what is being sent to the remove site is not the same as the same string I type in a browser.

Another easy example that works in a browser but not in my code:

```
   local http=require"httpm".create()
   print(lat,lon)
   local t,err=http:json("http://api.agify.io/?name=darren")
   print(t,err)

```
```
nil	nil
nil	Not Found
```

Thanks for any comments!

**Reply by surfskidude** [reply]
The documentation tells you to put the arguments in the data table.

* data - A table with key/value pairs (query data). This table may be empty.

`local t,err=http:json("http://api.sunrise-sunset.org/json",{lat=49,lng=-98,formatted=0})`

**Reply by darethehair** [reply]
Thanks so much for sending me in the right direction!  I was ultimately able to get all of my HTTP API-calling Lua code to work correctly because of this.  I still have a long ways to go, but I have more confidence that I will be able to do what I want.

**Comment by surfskidude** [comment]
If you are interested, I can send you a weather station example. The server side, which saves data in a time series file DB on the SD card, is more or less completed. However, the user interface, which should be designed as a SPA has not been designed.

**Reply by darethehair** [reply]
Sure, that would be cool to see :)

**Reply by surfskidude** [reply]
I created a new repo for the code: https://github.com/surfskidude/weather-station
